### DISCUSSION TITLE
Submitted values changed between calling debounced submit and actually executing the submit

### DISCUSSION THREAD

**Discussion by radoslavkarlik** [react-hook-form/react-hook-form#9047] [Q&A]
When you change the values (valid) then the success callback is debounced, when you change the values again (invalid) then no callback is called but the previously debounced callback will have the invalid form values.

How to reproduce: https://codesandbox.io/s/infallible-aj-oe1z7q?file=/src/App.js

The example has simplified submitting on form change.

When you type something and immediately remove it then the first 3 forms log the same values but the 4th form logs different values. It seems to be an issue when the form values are an array and a custom resolver is used.

**Comment by leapful** [comment]
@RadoslavK : You should debounce on the "onSubmit" callback of `<form>` element, not "onValid" of "handleSubmit" from "react-hook-form".

Working demo for SimpleWithoutCustomResolver component and ArrayWithCustomResolver component: https://codesandbox.io/s/relaxed-yonath-h783nq?file=/src/App.js

**Reply by radoslavkarlik** [reply]
Actually, I debounce the validation itself so it is not performed on every keystroke and then there is a longer success callback debounce. The example is just to make a point. Your working demo does not perform the success callback if I manage to make the form invalid meanwhile.

I am not sure why the results of the array fields with custom resolver are different from the rest. It also happens with nested object fields. [v7.13.0](https://github.com/react-hook-form/react-hook-form/releases/tag/v7.13.0) is the version that broke the previous behavior which was a while ago.

The workaround is to serialize or deep copy the form values before the debounced callback but I think it is not correct.

Use case for context:

- Form with an autosave with more complicated interaction and validation rules.
- The user changes something and the form is valid, the form is saved to the server after 1 second.
- If the user changes something again and the form becomes invalid before the saving request was fired, it is still fired with the previously correct values and the form still reflects its new invalid state that is not saved.

**Reply by leapful** [reply]
@RadoslavK : Form values from "handleSubmit" is a reference object values, therefore when using resolver, you need to clone it to avoid side effect being made on the old reference. The same handling is applied for 3rd party library like yup or zod when the validdation library will try to parse the form values object and validate it then return the new valid form values  related to the schema.

Working demo: https://codesandbox.io/s/floral-microservice-e1gq5x

**Reply by radoslavkarlik** [reply]
Thanks for pointing it out. Now I know why it actually happens and how it should be approached.
